pkg/datastore/lrdatastore.go

```diff
@@ -0,0 +1,159 @@
+	CacheBlock(ctx context.Context, block *structs.CompleteBlockstruct) error
+	GetMaxProfitHeader(ctx context.Context, slot uint64) (structs.HeaderAndTrace, error)
+
+	// to be changed
+	GetHeadersBySlot(ctx context.Context, slot uint64) ([]structs.HeaderAndTrace, error)
+	GetHeadersByBlockHash(ctx context.Context, hash types.Hash) ([]structs.HeaderAndTrace, error)
+	GetHeadersByBlockNum(ctx context.Context, num uint64) ([]structs.HeaderAndTrace, error)
+	GetLatestHeaders(ctx context.Context, limit uint64, stopLag uint64) ([]structs.HeaderAndTrace, error)
+	GetDeliveredBatch(context.Context, []structs.PayloadQuery) ([]structs.BidTraceWithTimestamp, error)
+
+	FixOrphanHeaders(context.Context, time.Duration) error
+	MemoryCleanup(context.Context, time.Duration, time.Duration) error
+}
+
+type RemoteDatastore interface {
+	GetPayload(context.Context, structs.PayloadKey) (*structs.BlockAndTrace, error)
+	PutPayload(context.Context, structs.PayloadKey, *structs.BlockAndTrace, time.Duration) error
+}
+
+type LocalRemoteDatastore struct {
+	LocalDatastore
+	Remote RemoteDatastore
+
+	Logger log.Logger
+
+	m LRDatastoreMetrics
+}
+
+func NewLocalRemoteDatastore(local LocalDatastore, remote RemoteDatastore, l log.Logger) *LocalRemoteDatastore {
+	s := LocalRemoteDatastore{
+		LocalDatastore: local,
+		Remote:         remote,
+		Logger:         l.WithField("relay-service", "stream-datastore"),
+	}
+
+	s.initMetrics()
+
+	return &s
+}
+
+type getPayloadResponse struct {
+	id        string
+	block     *structs.BlockAndTrace
+	isLocal   bool
+	fromCache bool
+	err       error
+}
+
+var chanPool = sync.Pool{
+	New: func() any {
+		return make(chan getPayloadResponse)
+	},
+}
+
+func (s *LocalRemoteDatastore) GetPayload(ctx context.Context, key structs.PayloadKey) (*structs.BlockAndTrace, bool, error) {
+	timer0 := prometheus.NewTimer(s.m.Timing.WithLabelValues("getPayload", "all"))
+	defer timer0.ObserveDuration()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	responses := chanPool.Get().(chan getPayloadResponse)
+	defer chanPool.Put(responses)
+
+	id := uuid.NewString()
+
+	go func(ctx context.Context, resp chan getPayloadResponse, id string) {
+		timer1 := prometheus.NewTimer(s.m.Timing.WithLabelValues("getPayload", "local"))
+		block, fromCache, err := s.LocalDatastore.GetPayload(ctx, key)
+		timer1.ObserveDuration()
+		select {
+		case responses <- getPayloadResponse{id: id, block: block, isLocal: true, fromCache: fromCache, err: err}:
+		case <-ctx.Done():
+		}
+	}(ctx, responses, id)
+
+	go func(ctx context.Context, resp chan getPayloadResponse, id string) {
+		timer1 := prometheus.NewTimer(s.m.Timing.WithLabelValues("getPayload", "remote"))
+		block, err := s.Remote.GetPayload(ctx, key)
+		timer1.ObserveDuration()
+		select {
+		case responses <- getPayloadResponse{id: id, block: block, isLocal: false, fromCache: false, err: err}:
+		case <-ctx.Done():
+		}
+	}(ctx, responses, id)
+
+	for i := 0; i < 2; i++ {
```

Comment:
I'm not sure you need any number here, just for simple for loop would be probably enough, as you would like to terminate it either on any successful compatible(id) return or ctx.Done() . You wouldn't need a counter as well